qiskit/synthesis/linear/linear_depth_lnn.py

```diff
@@ -0,0 +1,273 @@
+
+    return cx_instructions_rows
+
+
+def _north_west_to_identity(n, mat):
+    # Transform a north-west triangular matrix to identity in depth 3*n by Proposition 7.4 of [1]
+
+    # At start the labels are in reversed order
+    label_arr = list(reversed(range(n)))
+    first_qubit = 0
+    empty_layers = 0
+    done = False
+    cx_instructions_rows = []
+
+    while not done:
+        at_least_one_needed = False
+
+        for i in range(first_qubit, n - 1, 2):
+            # Exchange the labels if needed
+            if label_arr[i] > label_arr[i + 1]:
+                at_least_one_needed = True
+
+                # If row i has "1" in column i+1, swap and remove the "1" (in depth 2)
+                # otherwise, only do a swap (in depth 3)
+                if not mat[i, label_arr[i + 1]]:
+                    # Adding this turns the operation to a SWAP
+                    _row_op_update_instructions(cx_instructions_rows, mat, i + 1, i)
+
+                _row_op_update_instructions(cx_instructions_rows, mat, i, i + 1)
+                _row_op_update_instructions(cx_instructions_rows, mat, i + 1, i)
+
+                label_arr[i], label_arr[i + 1] = label_arr[i + 1], label_arr[i]
+
+        if not at_least_one_needed:
+            empty_layers += 1
+            if empty_layers > 1:  # if nothing happened twice in a row, then finished.
+                done = True
+        else:
+            empty_layers = 0
+
+        first_qubit = int(not first_qubit)
+
+    return cx_instructions_rows
+
+
+def _optimize_cx_circ_depth_5n_line(mat):
+    # Optimize CX circuit in depth bounded by 5n for LNN connectivity.
+    # The algorithm [1] has two steps:
+    # a) transform the originl matrix to a north-west matrix (m2nw),
+    # b) transform the north-west matrix to identity (nw2id).
+    #
+    # A square n-by-n matrix A is called north-west if A[i][j]=0 for all i+j>=n
+    # For example, the following matrix is north-west:
+    # [[0, 1, 0, 1]
+    #  [1, 1, 1, 0]
+    #  [0, 1, 0, 0]
+    #  [1, 0, 0, 0]]
+
+    # According to [1] the synthesis is done on the inverse matrix
+    # so the matrix mat is inverted at this step
+    mat_inv = deepcopy(mat)
+    mat_cpy = calc_inverse_matrix(mat_inv)
+
+    n = len(mat_cpy)
+
+    # Transform an arbitrary invertible matrix to a north-west triangular matrix
+    # by Proposition 7.3 of [1]
+    cx_instructions_rows_m2nw = _matrix_to_north_west(n, mat_cpy, mat_inv)
+
+    # Transform a north-west triangular matrix to identity in depth 3*n
+    # by Proposition 7.4 of [1]
+    cx_instructions_rows_nw2id = _north_west_to_identity(n, mat_cpy)
+
+    return cx_instructions_rows_m2nw, cx_instructions_rows_nw2id
+
+
+def synth_cnot_depth_line_kms(mat):
+    """
+    Synthesis algorithm for linear reversible circuits from [1], Chapter 7.
+    Synthesizes any linear reversible circuit of n qubits over linear nearest-neighbor
+    architecture using CX gates with depth at most 5*n.
```

Comment:
The first line of a docstring is more significant than others because it ends up being used as a summary line for the function and put it table of contents, etc. We should probably put something more descriptive here and save the reference for part of the docstring body. For example maybe something like:

```suggestion
    Synthesize linear reversible circuit for linear nearest-neighbor architectures using Kutin, Moulton, Smithline method.

    Synthesis algorithm for linear reversible circuits from [1], Chapter 7.
    Synthesizes any linear reversible circuit of n qubits over linear nearest-neighbor
    architecture using CX gates with depth at most 5*n.
```